### PR TITLE
feat(core,cart): extend and clarify DaffStates and DaffStateable

### PR DESCRIPTION
**Pull request by damienwebdev**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently, DaffState has details shared across several packages.

Fixes: N/A

## What is the new behavior?
This moves most of that burden to core for organizational purposes. I've also outlined a new interface called DaffStateable which would be implemented to enforce consistency across the codebase.

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

Users who depend on the old `DaffLoadingState` enum should use the new `DaffStates` enum.

## Other information

### CODE REVIEW

File: libs/cart/state/src/reducers/cart-billing-address/cart-billing-address.reducer.ts [hunk 1]
```diff
@@ -42,7 +42,7 @@ export function cartBillingAddressReducer<T extends DaffCart>(
           ...state.cart,
           billing_address: action.payload
         },
-        ...setLoading(state.loading, DaffLoadingState.Complete),
+        ...setLoading(state.loading, DaffStates.Complete),
```
Comment (griest024) [inline libs/cart/state/src/reducers/cart-billing-address/cart-billing-address.reducer.ts:45]: All of these completes should be stable instead, right? Are you going to do that in another PR?

File: libs/cart/state/src/reducers/cart-order/cart-order-state.interface.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
-import { DaffLoadingState, DaffStateError } from '@daffodil/core/state';
+import { DaffStates, DaffStateError } from '@daffodil/core/state';
 import { DaffCartOrderResult } from '@daffodil/cart';
 
 export interface DaffCartOrderReducerState<T extends DaffCartOrderResult = DaffCartOrderResult> {
   cartOrderResult: T;
-  loading: DaffLoadingState;
+  loading: DaffStates;
```
Comment (griest024) [inline libs/cart/state/src/reducers/cart-order/cart-order-state.interface.ts:6]: This can still be typed as a loading mutable state, right?

File: libs/core/state/src/states/loading.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { DaffStates } from "./state.enum";
```
Comment (griest024) [inline libs/core/state/src/states/loading.ts:1]: Use single quotes

File: libs/core/state/src/states/mutable-loading.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { DaffLoadingState } from "./loading";
+import { DaffStates } from "./state.enum";
+
+/**
+ * A description for an object that is only loadable, but changeable.
```
Comment (griest024) [inline libs/core/state/src/states/mutable-loading.ts:5]: I don't think only loadable makes sense here

File: libs/core/state/src/states/state.enum.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
+/**
+ * The DaffState enumerates the vast majority of state modifications that can exist 
+ * in UI applications. 
+ */
+export enum DaffStates {
```
Comment (griest024) [inline libs/core/state/src/states/state.enum.ts:5]: Is it DaffState singular or DaffStates plural?

File: libs/core/state/src/states/state.enum.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  /**
+   * The "Stable" state represents when no action is being taken upon an object.
+   */
+  Stable = 'Stable',
+
+  /**
+   * The "Resolving" state represents when there's uncertainty in the current status
```
Comment (griest024) [inline libs/core/state/src/states/state.enum.ts:12]: Does it represent uncertainty? I would say it's explicitly just for the immutable loading of some data, analogous to a classic GET request.

File: libs/core/state/src/states/stateable.interface.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { DaffStates } from './state.enum';
+
+/**
+ * The interface an object must implement to consistently describe its current 
+ * state.
+ */
+export interface DaffStateable<T extends DaffStates> {
```
Comment (griest024) [inline libs/core/state/src/states/stateable.interface.ts:7]: Would DaffLoadingState satisfy this generic constraint? I don't think it would.

**Review by griest024 — CHANGES_REQUESTED** [review]

**Review by lderrickable — CHANGES_REQUESTED** [review]

File: libs/cart/state/src/reducers/cart-billing-address/cart-billing-address.reducer.ts (hunk 1 above)
Comment (damienwebdev) [reply libs/cart/state/src/reducers/cart-billing-address/cart-billing-address.reducer.ts:45]: Yes. I'd like to leave this diff simply the transfer of DaffLoadingState to the equivalent DaffStates.

File: libs/core/state/src/states/state.enum.ts (hunk 5 above)
Comment (damienwebdev) [reply libs/core/state/src/states/state.enum.ts:5]: DaffState singular is better. Changed.

File: libs/core/state/src/states/mutable-loading.ts (hunk 4 above)
Comment (damienwebdev) [reply libs/core/state/src/states/mutable-loading.ts:5]: Fixed.

File: libs/core/state/src/states/loading.ts (hunk 3 above)
Comment (damienwebdev) [reply libs/core/state/src/states/loading.ts:1]: Resolved.

File: libs/cart/state/src/reducers/cart-order/cart-order-state.interface.ts (hunk 2 above)
Comment (damienwebdev) [reply libs/cart/state/src/reducers/cart-order/cart-order-state.interface.ts:6]: Yes.

File: libs/core/state/src/states/state.enum.ts (hunk 6 above)
Comment (damienwebdev) [reply libs/core/state/src/states/state.enum.ts:12]: I think we need to highlight that we're typically doing this when we don't already know the information. E.g. We're uncertain whether or not the content we're attempting to load may or may not exist.

File: libs/core/state/src/states/stateable.interface.ts (hunk 7 above)
Comment (damienwebdev) [reply libs/core/state/src/states/stateable.interface.ts:7]: Yes, it does.

File: libs/core/state/src/states/state.enum.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+   * of a state, e.g. when we're loading something that hasn't been loaded into 
+   * state before.
+   */
+  Resolving = 'Resolving',
+
+  /**
+   * The "Adding" state represents a temporary period while an item is currently
+   * in the process of being added to state.
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/core/state/src/states/state.enum.ts:20]: a20c9be6

**Comment by damienwebdev** [comment]
@lderrickable @griest024 I had to rebase the PR, please take a second look.

**Review by lderrickable — APPROVED** [review]
